src/Time/Formatting.hs

```diff
@@ -28,29 +33,30 @@ module Time.Formatting
        , unitsF
        ) where
 
-import Time.Rational (Rat, withRuntimeDivRat)
-import Time.Units (AllUnits, KnownRatName, Time, floorUnit, toUnit)
+import Time.Rational (withRuntimeDivRat)
+import Time.Units (AllTimes, KnownRatName, Time, floorUnit, toUnit)
 
 -- | Class for time formatting.
-class Series (units :: [Rat]) where
+class Series (times :: [*]) where
     seriesF :: forall unit . KnownRatName unit
             => Time unit
             -> String
 
-instance Series ('[] :: [Rat]) where
+instance Series ('[] :: [*]) where
     seriesF :: Time someUnit -> String
     seriesF _ = ""
 
-instance (KnownRatName unit, Series units) => Series (unit ': units :: [Rat]) where
-   seriesF :: forall someUnit . (KnownRatName someUnit)
-           => Time someUnit
-           -> String
-   seriesF t = let newUnit = withRuntimeDivRat @someUnit @unit $ toUnit @(Time unit) t
-                   format  = floorUnit newUnit
-                   timeStr = case (floor @(Time unit) newUnit :: Int) of
-                                  0 -> ""
-                                  _ -> show format
-               in timeStr ++ seriesF @units @unit (newUnit - format)
+instance (time ~ Time unit, KnownRatName unit, Series times)
+    => Series (time ': times :: [*]) where
+    seriesF :: forall someTime someUnit . (someTime ~ Time someUnit, KnownRatName someUnit)
+            => someTime
+            -> String
+    seriesF t = let newUnit = withRuntimeDivRat @someUnit @unit $ toUnit @(Time unit) t
+                    format  = floorUnit newUnit
+                    timeStr = case (floor @time newUnit :: Int) of
```

Comment:
I think `()` are redundant here. You can try to run `hlint` over our project. But I'm not sure it can parse it...